mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp

```diff
@@ -801,13 +801,31 @@ struct WarpOpTransferRead : public OpRewritePattern<WarpExecuteOnLane0Op> {
   using OpRewritePattern<WarpExecuteOnLane0Op>::OpRewritePattern;
   LogicalResult matchAndRewrite(WarpExecuteOnLane0Op warpOp,
                                 PatternRewriter &rewriter) const override {
-    OpOperand *operand = getWarpResult(
-        warpOp, [](Operation *op) { return isa<vector::TransferReadOp>(op); });
-    if (!operand)
-      return failure();
-    auto read = operand->get().getDefiningOp<vector::TransferReadOp>();
-    // Don't duplicate transfer_read ops when distributing.
-    if (!read.getResult().hasOneUse())
+    // Try to find a distributable yielded read. Note that this pattern can
+    // still fail at the end after distribution, in which case this might have
+    // missed another distributable read.
+    vector::TransferReadOp read;
+    auto yield = cast<vector::YieldOp>(
```

Comment:
This is mostly duplicated logic with `getWarpResult`. Should we create `getWarpResults` that return a list of candidates? I'd assume that'd be useful for handling other op cases too? If you don't want to go handle other op cases, we can change `getWarpResult` to just call `getWarpResults` for the first candidate to maintain its existing behavior.

Reply:
I realized I can just make the extra checks a part of the filter function; looks a lot cleaner now. There is still a problem where if either the lane fails to delinearize, or if there are leftover producers in the body of the warp op, it might miss out on propagating a different read, but by that point we've already cloned the warp op and attempted to move things over, so we can't really roll back and try again.

Reply:
Ha, good points!